### PR TITLE
3Ps

### PR DESCRIPTION
**Pull request by hellwolf**

### CODE REVIEW

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 1]
```diff
@@ -864,4 +895,40 @@ contract ConstantFlowAgreementV1 is
...
+    function _decode3PsData(
+        ISuperfluidToken token
+    ) 
+        internal view 
+        returns(uint256 liquidationPeriod, uint256 patricianPeriod) 
+    {
+        ISuperfluid host = ISuperfluid(token.getHost());
+        ISuperfluidGovernance gov = ISuperfluidGovernance(host.getGovernance());
+        uint256 configData = gov.getConfigAsUint256(host, token, _3PS_CONFIG_KEY);
+        liquidationPeriod = (configData >> 32) & type(uint32).max;
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:931]: changing of liquidation period for tokens w/ open streams

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 2]
```diff
@@ -747,33 +753,58 @@ contract ConstantFlowAgreementV1 is
...
         int256 signedSingleDeposit = flowData.deposit.toInt256();
         int256 signedTotalDeposit = senderAccountState.deposit.toInt256();
+        address bondAccount;
+        bool isPatricianPeriod;
+
+        // To handle stack too deep overflow issue
+        {
+            ISuperfluid host = ISuperfluid(token.getHost());
+            bondAccount = ISuperfluidGovernance(host.getGovernance())
+                .getConfigAsAddress(host, token, _REWARD_ADDRESS_CONFIG_KEY);
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:763]: should not read _REWARD_ADDRESS_CONFIG_KEY in cfa

File: packages/ethereum-contracts/contracts/superfluid/SuperfluidToken.sol [hunk 3]
```diff
@@ -407,6 +408,61 @@ abstract contract SuperfluidToken is ISuperfluidToken
             );
         }
     }
+
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/SuperfluidToken.sol:411]: add comments for parameters - natspec

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol [hunk 4]
```diff
@@ -11,6 +11,21 @@ import { ISuperAgreement } from "./ISuperAgreement.sol";
...
+     * @dev Encoded liquidation type data mainly used for handling stack to deep errors
+     * 
+     * Note:
+     * - version: 1 
+     * - liquidationType key:
+     *    - 0 = reward account receives reward (PIC period)
+     *    - 1 = liquidator account receives reward (Pleb period)
+     *    - 2 = liquidator account receives reward (Pirate period/bailout)
+     */
+    struct LiquidationTypeData {
```
[the remark this answers is not in the document]
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol:24]: after reviewing the natspec docs, structs do not require natspec docs: https://docs.soliditylang.org/en/v0.8.11/natspec-format.html#documentation-example

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 5]
```diff
@@ -746,34 +741,59 @@ contract ConstantFlowAgreementV1 is
...
-        int256 signedTotalDeposit = senderAccountState.deposit.toInt256();
+        // TODO: GDA deposit should be considered here too
+        int256 signedTotalCFADeposit = senderAccountState.deposit.toInt256();
+        bool isPatricianPeriod;
+
+        // To retrieve patrician period
+        // Note: curly brackets are to handle stack too deep overflow issue
+        {
+            if (senderAccountState.flowRate == 0) {
+            // if the flowRate is 0, it is always the patricianPeriod
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:752]: add comment here that this should never happen

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 6]
```diff
@@ -746,34 +741,59 @@ contract ConstantFlowAgreementV1 is
...
-            token.makeLiquidationPayouts(
-                flowParams.flowId,
-                liquidator,
-                flowParams.sender,
-                rewardAmount.toUint256(),
-                0
+            int256 rewardAmount = signedSingleDeposit.mul(totalRewardLeft).div(signedTotalCFADeposit);
+            token.makeLiquidationPayoutsV2(
+                flowParams.flowId, // id
+                abi.encode(1, isPatricianPeriod ? 0 : 1), // liquidationTypeData (1 means "v1") - 0 or 1 for pic or pleb
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:778]: `1 means "v1" of this encoding schema`

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol:24]: look into how to apply this to struct

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol [hunk 7]
```diff
@@ -349,6 +364,73 @@ interface ISuperfluidToken {
...
+     *   - the liquidatorAccount will get the rewardAmount (remaining deposit)
+     *   - the penaltyAccount will pay for the rewardAmount
+     * - if the penaltyAccount is insolvent
+     *   - the liquidatorAccount will get the rewardAmount (single deposit)
+     *   - the rewardAccount will pay for both the rewardAmount and bailoutAmount
+     *   - the penaltyAccount will receive the bailoutAmount
+     */
+    event AgreementLiquidatedByV2(
+        address liquidatorAccount,
+        address indexed agreementClass,
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol:393]: `agreementClass` then `liquidatorAccount`

Comment (hellwolf) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidToken.sol:393]:
```
    event AgreementLiquidatedV2(
        address indexed agreementClass,
        bytes32 id,
        address indexed liquidatorAccount,
        address indexed penaltyAccount,
        address indexed rewardAccount,
        uint256 rewardAmount,
        int256 penaltyAccountBalanceDelta,
        bytes liquidationTypeData
    );
```

**Review by hellwolf — COMMENTED** [review]
new format proposed for AgreementLiquidatedV2

File: packages/ethereum-contracts/contracts/superfluid/SuperfluidToken.sol [hunk 8]
```diff
@@ -407,6 +408,61 @@ abstract contract SuperfluidToken is ISuperfluidToken
...
+            // this can occur in two cases:
+            // - pleb period: the two amounts cancel each other out
+            // - pirate/bailout period: reward account has to pay bailout amount
+            _balances[rewardAccount] = _balances[rewardAccount]
+                .sub(rewardAmount.toInt256())
+                .sub(penaltyAccountBalanceDelta);
+        }
+
+        _balances[penaltyAccount] = _balances[penaltyAccount]
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/SuperfluidToken.sol:452]: `penaltyAccount` => `targetAccount`

Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/superfluid/SuperfluidToken.sol:452]: comments on balanceDelta + => bailout, else normal liquidation

**Comment by hellwolf** [comment]
closing, and to be recreated by @0xdavinchee

**Pull request closed without merging**